Decouple SOLVCON repository for devenv building

tai271828 opened the pull request:
By copying the files used to build devenv from SOLVCON repository, we
could build devenv without depending on SOLVCON.

Issue #6

tai271828 commented on prepare-solvcon-dev.sh:
```diff
@@ -45,11 +43,14 @@ conda update -q conda
-${SCSRC}/contrib/conda.sh
-${SCSRC}/contrib/build-pybind11-in-conda.sh
+# fetch SOLVCON source
+git clone https://github.com/solvcon/solvcon.git ${SCSRC}
```
SOLVCON repository is used for testing the devenv that has been built. The building process of devenv does not really need SOLVCON repository now.

yungyuc commented on conda.sh:
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+conda install -y \
+  python=3.6 \
```
solvcon and the associated projects should be upgraded to the latest Python releases, in the future.

tai271828 replied:
I got this idea and +1 to it. We may add option to choose python version in the future. I did not change anything for this option in this pull request this time.

yungyuc replied:
Thanks.  Comments like this are useful.  I saw your smiley but smileys are hard to track.

yungyuc commented on devenv-cli.sh:
```diff
@@ -1,4 +1,4 @@
-SOLVCON_WORKING_ROOT=${HOME}/solvcon/devenv
+SOLVCON_DEVENV_SRC=${HOME}/solvcon/devenv
```
The name is a little bit long for typing.  How about using `SCDE` (SOLVCON dev env) and making it `SCDE_SRC`?

tai271828 replied:
+1. Updated the PR.

tai271828 replied:
`SOLVCON_DEVENV_SRC` has been renamed as `SCDE`.

yungyuc commented on prepare-solvcon-dev.sh:
```diff
@@ -45,11 +43,14 @@ conda update -q conda
 conda create -p ${SCSRC_WORKING}/venv-conda --no-default-packages -y python
 source activate ${SCSRC_WORKING}/venv-conda
 
+# prepare all packages to build SOLVCON
```
LGTM

yungyuc commented:
@tai271828 Could you add some comments on the code that you want me to review?

tai271828 commented:
> @tai271828 Could you add some comments on the code that you want me to review?

Hi, I updated this pull request according to the previous comments above. If the comment is marked as "resolved", it means that 1) I updated the pull request according to the consensus if it is requested to fix 2) the code snippet does not need to update after we exchanged thoughts.

You may unfold the above "resolved" conversation blocks to trace the following change. Let me add more comments on them for more clear statement.

yungyuc commented:
> You may unfold the above "resolved" conversation blocks to trace the following change. Let me add more comments on them for more clear statement.

There are days between the activities of this PR and I have lost track of it.  Github code review seems hard to use.  It seems that the force-push with this PR worsens it, but I am not totally sure.

I guess I already lost track when you [requested my review 7 days ago](https://github.com/solvcon/devenv/pull/8#event-4270413068).  Adding comments to indicate what to be reviewed and avoid force push would help.

yungyuc commented:
I think this PR didn't completely decouple SOLVCON from devenv.  We will need follow-up work.

tai271828 commented:
> > You may unfold the above "resolved" conversation blocks to trace the following change. Let me add more comments on them for more clear statement.
> 
> There are days between the activities of this PR and I have lost track of it. Github code review seems hard to use. It seems that the force-push with this PR worsens it, but I am not totally sure.
> 
> I guess I already lost track when you [requested my review 7 days ago](https://github.com/solvcon/devenv/pull/8#event-4270413068). Adding comments to indicate what to be reviewed and avoid force push would help.

If we are happy to include trivial commits like "fix typo" or "rename variable", I could stop using force-push to a pull request. Let's try next time?

tai271828 commented:
> I think this PR didn't completely decouple SOLVCON from devenv. We will need follow-up work.

https://github.com/solvcon/devenv/issues/6 is not yet closed. We could use https://github.com/solvcon/devenv/issues/6 for follow-up work.

yungyuc commented:
> > > You may unfold the above "resolved" conversation blocks to trace the following change. Let me add more comments on them for more clear statement.
> > 
> > 
> > There are days between the activities of this PR and I have lost track of it. Github code review seems hard to use. It seems that the force-push with this PR worsens it, but I am not totally sure.
> > I guess I already lost track when you [requested my review 7 days ago](https://github.com/solvcon/devenv/pull/8#event-4270413068). Adding comments to indicate what to be reviewed and avoid force push would help.
> 
> If we are happy to include trivial commits like "fix typo" or "rename variable", I could stop using force-push to a pull request. Let's try next time?

We should avoid trivial commits to the PR branch too.

I think a better way is to use two branches for development.  We start with a development branch, e.g., `wip/something`. 
 Once the code is ready for PR/review, we create `feature/something`, merge `wip/something` to it, and file PR with `feature/something`.  Subsequent development and trivial commits happen in `wip/something`.  Once the code is ready for review, it's `wip/something` is merged to `feature/something`.

The merge from `wip/something` to `feature/something` may be a squash, or the merge should contain sufficient checkin log, unlike a normal merge.  I guess squash might work better, but am not totally certain.  `wip/something` may force-push.

tai271828 commented:
> > > > You may unfold the above "resolved" conversation blocks to trace the following change. Let me add more comments on them for more clear statement.
> > > 
> > > 
> > > There are days between the activities of this PR and I have lost track of it. Github code review seems hard to use. It seems that the force-push with this PR worsens it, but I am not totally sure.
> > > I guess I already lost track when you [requested my review 7 days ago](https://github.com/solvcon/devenv/pull/8#event-4270413068). Adding comments to indicate what to be reviewed and avoid force push would help.
> > 
> > 
> > If we are happy to include trivial commits like "fix typo" or "rename variable", I could stop using force-push to a pull request. Let's try next time?
> 
> We should avoid trivial commits to the PR branch too.
> 
> I think a better way is to use two branches for development. We start with a development branch, e.g., `wip/something`.
> Once the code is ready for PR/review, we create `feature/something`, merge `wip/something` to it, and file PR with `feature/something`. Subsequent development and trivial commits happen in `wip/something`. Once the code is ready for review, it's `wip/something` is merged to `feature/something`.
> 
> The merge from `wip/something` to `feature/something` may be a squash, or the merge should contain sufficient checkin log, unlike a normal merge. I guess squash might work better, but am not totally certain. `wip/something` may force-push.

I use this two-branch flow very often, and like it very much. What if we create a pull request with `wip/something` next time, and create another PR of `feature/something` based on a (rebased and reviewed) `wip/something`?

FYI, Github offers "convert to draft" feature for each pull request, but this feature is not useful if we want to exclude trivial commits without force-push to the same pull request.

yungyuc commented:
Sounds good.  I created a [wiki page](https://github.com/solvcon/devenv/wiki/Branching-flow) for the branching flow.  Feel free to edit.
